src/main/java/net/neoforged/neoforge/energy/ComponentEnergyStorage.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.energy;
+
+import net.minecraft.core.component.DataComponentType;
+import net.minecraft.util.Mth;
+import net.minecraft.world.item.ItemStack;
+import net.neoforged.neoforge.capabilities.ICapabilityProvider;
+import net.neoforged.neoforge.capabilities.RegisterCapabilitiesEvent;
+import net.neoforged.neoforge.common.MutableDataComponentHolder;
+
+/**
+ * Variant of {@link EnergyStorage} for use with data components.
+ * <p>
+ * The actual data storage is managed by a data component, and all changes will write back to that component.
+ * <p>
+ * To use this class, register a new {@link DataComponentType} which holds an {@link Integer} for your item.
+ * Then reference that component from your {@link ICapabilityProvider} passed to {@link RegisterCapabilitiesEvent#registerItem} to create an instance of this class.
+ */
+public class ComponentEnergyStorage implements IEnergyStorage {
+    protected final MutableDataComponentHolder parent;
+    protected final DataComponentType<Integer> energyComponent;
```

Comment:
Do we want to change this to a `Supplier<DataComponentType<Integer>>`?

Reply:
I don't think there's any benefit in doing that, it just moves the .get() call from the consumer (at the ctor) to every field access

Reply:
Just asking for consistency with `FluidHandlerItemStack`.

Reply:
Electing to proceed with the non-supplier options per https://discord.com/channels/313125603924639766/1154167065519861831/1244080298795335721

It wouldn't make sense to use the supplier when we need to immediately resolve it for sanity checking.